core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -136,30 +169,33 @@ private boolean checkTasks() {
         }
       }
 
-      return !closed && (tasks.hasNext() || hasRunningTask);
+      return !closed.get() && (tasks.hasNext() || hasRunningTask);
     }
 
-    private Future<?> submitNextTask() {
-      if (!closed && tasks.hasNext()) {
-        return workerPool.submit(tasks.next());
+    private CompletableFuture<Optional<Task<T>>> submitNextTask() {
+      if (!closed.get()) {
+        if (!yieldedTasks.isEmpty()) {
+          return CompletableFuture.supplyAsync(yieldedTasks.removeFirst(), workerPool);
+        } else if (tasks.hasNext()) {
+          return CompletableFuture.supplyAsync(tasks.next(), workerPool);
+        }
       }
       return null;
     }
 
     @Override
     public synchronized boolean hasNext() {
-      Preconditions.checkState(!closed, "Already closed");
-
-      // if the consumer is processing records more slowly than the producers, then this check will
-      // prevent tasks from being submitted. while the producers are running, this will always
-      // return here before running checkTasks. when enough of the tasks are finished that the
-      // consumer catches up, then lots of new tasks will be submitted at once. this behavior is
-      // okay because it ensures that records are not stacking up waiting to be consumed and taking
-      // up memory.
-      //
-      // consumers that process results quickly will periodically exhaust the queue and submit new
-      // tasks when checkTasks runs. fast consumers should not be delayed.
-      if (!queue.isEmpty()) {
+      Preconditions.checkState(!closed.get(), "Already closed");
+
+      // If the consumer is processing records more slowly than the producers, the producers will
+      // eventually fill the queue and yield, returning continuations. Continuations and new tasks
+      // are started by checkTasks(). The check here prevents us from restarting continuations or
+      // starting new tasks too early (when queue is almost full) or too late (when queue is already
+      // emptied). Restarting too early would lead to tasks yielding very quickly (CPU waste on
+      // scheduling). Restarting too late would mean the consumer may need to wait for the tasks
+      // to produce new items. A consumer slower than producers shouldn't need to wait.
+      int queueLowWaterMark = maxQueueSize / 2;
```

Comment:
Why is this a local variable? if it is a constant it could be a final field.

Reply:
It's a variable, so it has a name, which is aimed at help understanding what's going on.
maxQueueSize is a configuration parameter.  queueLowWaterMark can indeed be a field, i am not convinced it would increase readability, because the usage and the value formulae would be separated.

Reply:
I think either way doesn't affect readability. but since it is static, it might be a little bit to set in the constructor as a final field.

Reply:
technically, it's not static. this is a constructor parameter. the static field serves as a default.

Reply:
static is probably the wrong world. it is a constant for the object and could be a class field just like `maxQueueSize`

Reply:
re this being class static field -- currently this is instance parameter, the constant serves only as a default
re this being class non-static (instance) field -- i replied to this in https://github.com/apache/iceberg/pull/10691#discussion_r1684130747. let me know what you think about this.

Reply:
@findepi, I think by "static" Steven meant "does not change" rather than Java's `static`. This is minor, but we don't typically want to recalculate a value in every method call if it can be set when the instance is created. The name isn't the problem, it is that this doesn't need to be recalculated every call.